### PR TITLE
feat: Smart contract details page hints

### LINKED ISSUE #50: Smart contract details page hints
Display hints on contract page details.

### PR DESCRIPTION
**Pull request by janmichek**
## Description
resolves #50 

## Checklist:

- [x] I have read and followed the [Contributing Guide](../../CONTRIBUTING.md)  
- [x] add smart contract details page hints

### CODE REVIEW

**Comment by janmichek** [comment]
> @janmichek would it be better to solve conflicts first or can we review it?

Sure, it was hanging here for a while. Resolved

File: src/components/ContractDetailsPanel.vue [hunk 1]
```diff
@@ -133,6 +157,8 @@ import CopyChip from '@/components/CopyChip'
 import AppChip from '@/components/AppChip'
 import { formatAePrice, formatAettosToAe, formatEllipseHash } from '@/utils/format'
 import DatetimeLabel from '@/components/DatetimeLabel'
+import { contractsHints } from '~/utils/hints/contractsHints'
```
Comment (lukeromanowicz) [inline src/components/ContractDetailsPanel.vue:160]: I think we are using the wrong alias here and in other places

File: src/components/ContractEventsTable.vue [hunk 2]
```diff
@@ -38,6 +58,7 @@
 </template>
 
 <script setup>
+import { contractsHints } from '../utils/hints/contractsHints'
```
Comment (lukeromanowicz) [inline src/components/ContractEventsTable.vue:61]: would be nice to use @ alias everywhere

File: src/utils/hints/contractsHints.js [hunk 3]
```diff
@@ -3,4 +3,17 @@ export const contractsHints = {
...
+  contractCalls: 'Amount of calls to the Smart Contract.',
+  apiLinks: 'Links to sources of the Smart Contract data.',
+  callTransactionsHash: 'Unique identifier of ContractCallTx which represents call to Smart Contract function.',
+  callTransactionsTime: 'Estimated date, time and block height when the Smart Contract was called.',
+  callTransactionsCaller: 'Account that called the Smart Contract.',
+  callTransactionsEntryPoint: 'Name of the Smart Contract function that was called.',
+  eventsCallTransaction: 'Unique identifier of ContractCallTx which emitted the event.',
+  eventsCreated: 'Estimated date, time and block height when the Smart Contract was emitted the event.',
+  eventsName: 'Name of the emitted event',
+  eventsArguments: 'Payload emitted with the event.',
```
Comment (lukeromanowicz) [inline src/utils/hints/contractsHints.js:18]: What do you think about "The payload of the emitted event."?

File: src/utils/hints/contractsHints.js [hunk 4]
```diff
@@ -3,4 +3,17 @@ export const contractsHints = {
...
+  contractsAccountBalance: 'Smart Contract account balance. Balance is needed to execute "payable" functions.',
+  contractCalls: 'Amount of calls to the Smart Contract.',
+  apiLinks: 'Links to sources of the Smart Contract data.',
+  callTransactionsHash: 'Unique identifier of ContractCallTx which represents call to Smart Contract function.',
+  callTransactionsTime: 'Estimated date, time and block height when the Smart Contract was called.',
+  callTransactionsCaller: 'Account that called the Smart Contract.',
+  callTransactionsEntryPoint: 'Name of the Smart Contract function that was called.',
+  eventsCallTransaction: 'Unique identifier of ContractCallTx which emitted the event.',
+  eventsCreated: 'Estimated date, time and block height when the Smart Contract was emitted the event.',
+  eventsName: 'Name of the emitted event',
```
Comment (lukeromanowicz) [inline src/utils/hints/contractsHints.js:17]: Missing dot at the end of the sentence

File: src/utils/hints/contractsHints.js [hunk 5]
```diff
@@ -3,4 +3,17 @@ export const contractsHints = {
...
+  contractsAccount: 'The address where the Smart Contract is located. To be used if a Smart Contract is already deployed.',
+  contractsAccountBalance: 'Smart Contract account balance. Balance is needed to execute "payable" functions.',
+  contractCalls: 'Amount of calls to the Smart Contract.',
+  apiLinks: 'Links to sources of the Smart Contract data.',
+  callTransactionsHash: 'Unique identifier of ContractCallTx which represents call to Smart Contract function.',
+  callTransactionsTime: 'Estimated date, time and block height when the Smart Contract was called.',
+  callTransactionsCaller: 'Account that called the Smart Contract.',
+  callTransactionsEntryPoint: 'Name of the Smart Contract function that was called.',
+  eventsCallTransaction: 'Unique identifier of ContractCallTx which emitted the event.',
+  eventsCreated: 'Estimated date, time and block height when the Smart Contract was emitted the event.',
```
Comment (lukeromanowicz) [inline src/utils/hints/contractsHints.js:16]: Did you mean "(...) when the Smart Contract had the event  emitted."?

**Review by lukeromanowicz — CHANGES_REQUESTED** [review]
Looks pretty decent. I have only a few small notes to take a look at

File: src/components/ContractDetailsPanel.vue (hunk 1 above)
Comment (janmichek) [reply src/components/ContractDetailsPanel.vue:160]: fixed

File: src/components/ContractEventsTable.vue (hunk 2 above)
Comment (janmichek) [reply src/components/ContractEventsTable.vue:61]: fixed

File: src/utils/hints/contractsHints.js (hunk 3 above)
Comment (janmichek) [reply src/utils/hints/contractsHints.js:18]: accepted

File: src/utils/hints/contractsHints.js (hunk 4 above)
Comment (janmichek) [reply src/utils/hints/contractsHints.js:17]: fixed

File: src/utils/hints/contractsHints.js (hunk 5 above)
Comment (janmichek) [reply src/utils/hints/contractsHints.js:16]: accepted with a bit of rephrasal

Comment (lukeromanowicz) [reply src/utils/hints/contractsHints.js]: Thanks, even better!

**Review by lukeromanowicz — APPROVED** [review]
Thanks for the improvements, it looks spot on!

File: src/utils/hints/contractsHints.js [hunk 6]
```diff
@@ -3,4 +3,17 @@ export const contractsHints = {
   contractCreated: 'Keyblock height and exact date and time when the smart contract was created.',
   contractTxHash: 'The transaction hash of a ContractCreateTx that was executed to create the smart contract on the blockchain.',
   contractCreator: 'Account that created the smart contract.',
+  bytecode: 'Pre-compiled code of the smart contract.',
+  contractsAccount: 'The address where the smart contract is located. To be used if a smart contract is already deployed.',
+  contractsAccountBalance: 'smart contract account balance. Balance is needed to execute "payable" functions.',
```
Comment (michele-franchi) [inline src/utils/hints/contractsHints.js:8]:
```suggestion
  contractsAccountBalance: 'Smart contract account balance that is needed to execute "payable" functions.',
```

File: src/utils/hints/contractsHints.js [hunk 7]
```diff
@@ -3,4 +3,17 @@ export const contractsHints = {
   contractCreated: 'Keyblock height and exact date and time when the smart contract was created.',
   contractTxHash: 'The transaction hash of a ContractCreateTx that was executed to create the smart contract on the blockchain.',
   contractCreator: 'Account that created the smart contract.',
+  bytecode: 'Pre-compiled code of the smart contract.',
```
Comment (michele-franchi) [inline src/utils/hints/contractsHints.js:6]:
```suggestion
  bytecode: 'Low-level representation of the smart contract that is generated when the contract's source code is compiled.',
```

File: src/utils/hints/contractsHints.js [hunk 8]
```diff
@@ -3,4 +3,17 @@ export const contractsHints = {
...
+  contractsAccount: 'The address where the smart contract is located. To be used if a smart contract is already deployed.',
+  contractsAccountBalance: 'smart contract account balance. Balance is needed to execute "payable" functions.',
+  contractCalls: 'Amount of calls to the smart contract.',
+  apiLinks: 'Links to sources of the smart contract data.',
+  callTransactionsHash: 'Unique identifier of ContractCallTx which represents call to smart contract function.',
```
Comment (michele-franchi) [inline src/utils/hints/contractsHints.js:11]:
```suggestion
  callTransactionsHash: 'Unique identifier of the ContractCallTx transaction.',
```

File: src/utils/hints/contractsHints.js [hunk 9]
```diff
@@ -3,4 +3,17 @@ export const contractsHints = {
...
+  callTransactionsTime: 'Estimated date, time and block height when the smart contract was called.',
+  callTransactionsCaller: 'Account that called the smart contract.',
+  callTransactionsEntryPoint: 'Name of the smart contract function that was called.',
+  eventsCallTransaction: 'Unique identifier of ContractCallTx which emitted the event.',
```
Comment (michele-franchi) [inline src/utils/hints/contractsHints.js:15]:
```suggestion
  eventsCallTransaction: 'Unique identifier of the ContractCallTx transaction which emitted the event.',
```

**Comment by michele-franchi** [comment]
Once reviewed @janmichek I would suggest to also add @marc0olo for a quick review here. 

@marc0olo we are approaching the end of this 😃

**Comment by marc0olo** [comment]
@michele-franchi let me know when I should start checking. not sure if I should already have a look now

File: src/utils/hints/contractsHints.js (hunk 6 above)
Comment (janmichek) [reply src/utils/hints/contractsHints.js:8]: fixed

File: src/utils/hints/contractsHints.js (hunk 7 above)
Comment (janmichek) [reply src/utils/hints/contractsHints.js]: fixed

File: src/utils/hints/contractsHints.js (hunk 8 above)
Comment (janmichek) [reply src/utils/hints/contractsHints.js]: fixed

File: src/utils/hints/contractsHints.js (hunk 9 above)
Comment (janmichek) [reply src/utils/hints/contractsHints.js]: fixed

File: src/utils/hints/contractsHints.js [hunk 10]
```diff
@@ -3,4 +3,17 @@ export const contractsHints = {
...
+  eventsCreated: 'Estimated date, time and block height when the smart contract had emitted the event.',
+  eventsName: 'Name of the emitted event.',
+  eventsArguments: 'The payload of the emitted event.',
```
[the remark this answers is not in the document]
Comment (janmichek) [reply src/utils/hints/contractsHints.js]: fixed

**Comment by janmichek** [comment]
> Good job, I added some suggestions.
> 
> This table should be fixed (show decoded events) but in any case right now the "Data" tooltip is missing: image

The response in `data` attribute is actually empty `""`

**Comment by janmichek** [comment]
@marc0olo You can go now .)

File: src/components/ContractCallTransactionsTable.vue [hunk 11]
```diff
@@ -2,10 +2,30 @@
...
+          </hint-tooltip>
+        </th>
+        <th>
+          Caller
+          <hint-tooltip>
+            {{ contractsHints.callTransactionsCaller }}
+          </hint-tooltip>
+        </th>
+        <th>
+          Entry Point
```
Comment (marc0olo) [inline src/components/ContractCallTransactionsTable.vue:24]:
```suggestion
          Entrypoint
```

File: src/components/ContractCallTransactionsTableCondensed.vue [hunk 12]
```diff
@@ -37,7 +52,12 @@
       </tr>
       <tr class="contract-call-transactions-table-condensed__row">
         <th class="contract-call-transactions-table-condensed__header">
-          Entry Point
+          <app-tooltip>
+            Entry Point
```
Comment (marc0olo) [inline src/components/ContractCallTransactionsTableCondensed.vue:56]:
```suggestion
            Entrypoint
```

File: src/utils/hints/contractsHints.js [hunk 13]
```diff
@@ -3,4 +3,18 @@ export const contractsHints = {
   contractCreated: 'Keyblock height and exact date and time when the smart contract was created.',
   contractTxHash: 'The transaction hash of a ContractCreateTx that was executed to create the smart contract on the blockchain.',
   contractCreator: 'Account that created the smart contract.',
+  bytecode: 'Low-level representation of the smart contract that is generated when the contract\'s source code is compiled.',
+  contractsAccount: 'The address where the smart contract is located. To be used if a smart contract is already deployed.',
```
Comment (marc0olo) [inline src/utils/hints/contractsHints.js:7]:
```suggestion
  contractsAccount: 'The account associated to the smart contract.',
```

File: src/utils/hints/contractsHints.js [hunk 14]
```diff
@@ -3,4 +3,18 @@ export const contractsHints = {
...
+  contractsAccountBalance: 'Smart contract account balance that is needed to execute "payable" functions.',
+  contractCalls: 'Amount of calls to the smart contract.',
+  apiLinks: 'Links to sources of the smart contract data.',
+  callTransactionsHash: 'Low-level representation of the smart contract that is generated when the contract\'s source code is compiled.',
```
Comment (marc0olo) [inline src/utils/hints/contractsHints.js:11]:
```suggestion
  callTransactionsHash: 'Unique identifier of the ContractCallTx.',
```

File: src/utils/hints/contractsHints.js [hunk 15]
```diff
@@ -3,4 +3,18 @@ export const contractsHints = {
...
+  callTransactionsTime: 'Estimated date, time and block height when the smart contract was called.',
```
Comment (marc0olo) [inline src/utils/hints/contractsHints.js:12]:
```suggestion
  callTransactionsTime: 'Estimated date, time and keyblock height when the smart contract was called.',
```

is it estimated or exact date? also it seems like "estimated" also relates to the keyblock height which reads weird 🤔 please compare with other statements, I think we have multiple places for this

File: src/utils/hints/contractsHints.js [hunk 16]
```diff
@@ -3,4 +3,18 @@ export const contractsHints = {
...
+  callTransactionsCaller: 'Account that called the smart contract.',
+  callTransactionsEntryPoint: 'Name of the smart contract function that was called.',
```
Comment (marc0olo) [inline src/utils/hints/contractsHints.js:14]:
```suggestion
  callTransactionsEntryPoint: 'Name of the smart contract entrypoint that was called.',
```

Note: try to keep "function" and "entrypoint" generally consistent

File: src/utils/hints/contractsHints.js [hunk 17]
```diff
@@ -3,4 +3,18 @@ export const contractsHints = {
...
+  eventsCallTransaction: 'Unique identifier of the ContractCallTx transaction which emitted the event.',
```
Comment (marc0olo) [inline src/utils/hints/contractsHints.js:15]:
```suggestion
  eventsCallTransaction: 'Unique identifier of the ContractCallTx which emitted the event.',
```

File: src/utils/hints/contractsHints.js [hunk 18]
```diff
@@ -3,4 +3,18 @@ export const contractsHints = {
...
+  eventsCreated: 'Estimated date, time and block height when the smart contract had emitted the event.',
```
Comment (marc0olo) [inline src/utils/hints/contractsHints.js:16]: same as for `callTransactionsTime`

File: src/utils/hints/contractsHints.js [hunk 19]
```diff
@@ -3,4 +3,18 @@ export const contractsHints = {
...
+  eventsName: 'Name of the emitted event.',
+  eventsArguments: 'The arguments of the emitted event.',
+  data: 'The content of the emitted event.',
```
Comment (marc0olo) [inline src/utils/hints/contractsHints.js:19]: is data really relevant / used ? 🤔  aren't arguments the data emitted?

File: src/utils/hints/contractsHints.js [hunk 20]
```diff
@@ -3,4 +3,18 @@ export const contractsHints = {
...
+  contractCalls: 'Amount of calls to the smart contract.',
+  apiLinks: 'Links to sources of the smart contract data.',
+  callTransactionsHash: 'Low-level representation of the smart contract that is generated when the contract\'s source code is compiled.',
+  callTransactionsTime: 'Estimated date, time and block height when the smart contract was called.',
+  callTransactionsCaller: 'Account that called the smart contract.',
+  callTransactionsEntryPoint: 'Name of the smart contract function that was called.',
+  eventsCallTransaction: 'Unique identifier of the ContractCallTx transaction which emitted the event.',
+  eventsCreated: 'Estimated date, time and block height when the smart contract had emitted the event.',
+  eventsName: 'Name of the emitted event.',
+  eventsArguments: 'The arguments of the emitted event.',
```
Comment (marc0olo) [inline src/utils/hints/contractsHints.js:18]: I guess these are the values (data) emitted for the event? what format is this?

**Review by marc0olo — CHANGES_REQUESTED** [review]

File: src/utils/hints/contractsHints.js (hunk 20 above)
Comment (janmichek) [reply src/utils/hints/contractsHints.js:18]: @michele-franchi  ?

File: src/utils/hints/contractsHints.js (hunk 19 above)
Comment (janmichek) [reply src/utils/hints/contractsHints.js:19]: Afaig arguments ar data (parameters) passed to entrypoint

File: src/utils/hints/contractsHints.js (hunk 18 above)
Comment (janmichek) [reply src/utils/hints/contractsHints.js:16]: here it is estimated, but I rephrased the sentence

File: src/utils/hints/contractsHints.js (hunk 17 above)
Comment (janmichek) [reply src/utils/hints/contractsHints.js:15]: fixed

File: src/utils/hints/contractsHints.js (hunk 16 above)
Comment (janmichek) [reply src/utils/hints/contractsHints.js:14]: fixed

File: src/utils/hints/contractsHints.js (hunk 15 above)
Comment (janmichek) [reply src/utils/hints/contractsHints.js:12]: you are right, the time is exact.
I rephrased the sentence the same as in previous case

File: src/utils/hints/contractsHints.js (hunk 14 above)
Comment (janmichek) [reply src/utils/hints/contractsHints.js:11]: fixed

File: src/utils/hints/contractsHints.js [hunk 21]
```diff
@@ -3,4 +3,18 @@ export const contractsHints = {
   contractCreated: 'Keyblock height and exact date and time when the smart contract was created.',
   contractTxHash: 'The transaction hash of a ContractCreateTx that was executed to create the smart contract on the blockchain.',
   contractCreator: 'Account that created the smart contract.',
+  bytecode: 'Low-level representation of the smart contract that is generated when the contract\'s source code is compiled.',
+  contractsAccount: 'The address where the smart contract is located. To be used if a smart contract is already deployed.',
+  contractsAccountBalance: 'Smart contract account balance that is needed to execute "payable" functions.',
```
[the remark this answers is not in the document]
Comment (janmichek) [reply src/utils/hints/contractsHints.js:8]: fixed

File: src/utils/hints/contractsHints.js (hunk 13 above)
Comment (janmichek) [reply src/utils/hints/contractsHints.js:7]: fixed

File: src/components/ContractCallTransactionsTableCondensed.vue (hunk 12 above)
Comment (janmichek) [reply src/components/ContractCallTransactionsTableCondensed.vue:56]: same

File: src/components/ContractCallTransactionsTable.vue (hunk 11 above)
Comment (janmichek) [reply src/components/ContractCallTransactionsTable.vue:24]: Ok, as discussed before I have fixed this in previous PR, but also go aligned in this PR all over the app

File: src/utils/hints/contractsHints.js (hunk 20 above)
Comment (michele-franchi) [reply src/utils/hints/contractsHints.js:18]: Not sure if I got your question but these are encoded arguments.

Comment (marc0olo) [reply src/utils/hints/contractsHints.js:18]: to me this is a bit strange displayed. I assume these can be different (encoded) values. I am also not sure if I would replace "arguments" with "values". right now it is not clear if this is only one value or whether these can be different values. or is it one combined encoded value for all possible values/arguments ? 😅 

the thing is that it is cutted and can be copied, so the user doesn't really see the value(s). also, they are quite meaningless anyway at this point of time.

**Review by CharisSiarampalis — APPROVED** [review]

File: src/utils/hints/contractsHints.js (hunk 20 above)
Comment (michele-franchi) [reply src/utils/hints/contractsHints.js:18]: Well, I clarified this with Rogerio and I would do the following:
- Get rid of the "Data" column as it's a really specific use case
- Rename arguments column to "Data" to be consistent with [Token events](https://aescan.dev.aepps.com/tokens/ct_KeTvHnhU85vuuQMMZocaiYkPL9tkoavDRT3Jsy47LK2YqLHYb)

In a separate issue:
- Decode known args (using the parameter aexn-args=true)
- Find a better way to print arguments that cannot be decoded? In my opinion in any case these should be printed somehow and explain to the user what should be done in order to decode them

Comment (janmichek) [reply src/utils/hints/contractsHints.js:18]: @michele-franchi what you mean by `to be consistent with Token events` ? This issue is about Token events.

Comment (janmichek) [reply src/utils/hints/contractsHints.js:18]: Ah, sorry this is about contract events, got it .) Will fix it

Comment (janmichek) [reply src/utils/hints/contractsHints.js]: fixed, please check now

Comment (janmichek) [reply src/utils/hints/contractsHints.js]: Followup https://github.com/aeternity/aescan/issues/362

**Pull request merged**